### DISCUSSION TITLE
Best way to stream to another application

### DISCUSSION THREAD

**Discussion by danbo** [savonet/liquidsoap#2757] [User Support]
# Goal

Similarly to how LS can stream to output.icecast(), I would like to stream my source to an external application which will handle playing audio. I would also like LS to gracefully handle itself if the external application becomes unavailable, ie it does not stop/exit, output to other/existing sources is unaffected.

The external application currently accepts the following interfaces

1. pipes
2. unix sockets
3. tcp sockets
4. custom process

Here is what I've tried;

### 1a - output.file(pipe)

```
output.file(
  fallible=true,
  reopen_on_error=true,
  %wav,
  "/tmp/mypipe",
  s,
)
```
#### Result:
1. LS quits (despite having reopen_on_error=true) when the target application is stopped / stops reading the pipe. Didn't matter if I created a normal pipe or a non-blocking pipe.

### 1b - output.external(ffmpeg -> pipe)

```
output.external(
  reopen_on_error=true,
  %wav,
  "ffmpeg -y -i pipe:0 -f u16le -acodec pcm_s16le -ac 2 -ar 44100 /tmp/mypipe",
  s
)
```
#### Result:
1. Receiving "We must catchup (0.25-0.58) seconds! ..." (over 50 samples)
2. LS hangs when the target application is stopped / stops reading the pipe.
3. Any existing outputs are stalled, ie icecast
4a. if the pipe was created using mkfifo, everything automatically starts working again once the target application is back up.
4b. if the pipe was created using socat (removed when stopped), LS continues to hang and does not recover once the target application is back up.

### 2a - output.file(unix socket) 

```
output.file(
  reopen_on_error=true,
  %wav,
  "/tmp/my.sock",
  s,
)
```
#### Result: 
1. LS shuts down with "[clock.main:2] Source unix:/tmp/my.sock failed while streaming: Permission denied in mkdir(unix:)!" despite the socket being owned by the liquidsoap user with a+rwx permissions. Doesn't matter if I have the unix:// prefix on the socket path.

### 2b - output.external(ffmpeg -> unix socket)

```
output.external(
  reopen_on_error=true,
  %wav,
  "ffmpeg -y -i pipe:0 -f u16le -acodec pcm_s16le -ac 2 -ar 44100 unix:/tmp/my.sock",
  s
)
```
#### Result: 
1. LS loops to try and connect to unix socket disregarding reopen_delay 
2. Any existing output ie icecast appears to work / continue streaming
3. When external application becomes available, LS automatically starts streaming to socket again

### 3a - output.file(tcp socket)

```
output.file(
  reopen_on_error=true,
  %wav,
  "tcp://1.2.3.4:5678",
  s,
)
```
#### Result: 
1. Same behaviour as 2a

### 3b - output.external(ffmpeg -> tcp socket)

```
output.external(
  reopen_on_error=true,
  %wav,
  "ffmpeg -y -i pipe:0 -f u16le -acodec pcm_s16le -ac 2 -ar 44100 tcp://1.2.3.4:5678",
  s
)
```
#### Result: 
1. Same behaviour as 2b - "[clock.main:2] Source tcp://#{host}:#{port} failed while streaming: Permission denied in mkdir(tcp:)!"

## 4 - custom process

1. Have not tried this yet
2. Another option is srt - the external app doesn't support it but does support external tools like LS, ie ffmpeg, but it doesn't look like srt support is standard so I would have to compile it in.

# Summary

1. For now I will use sockets via ffmpeg because it works and doesn't stall LS or make it exit. 

```
output.external(
  reopen_on_error=true,
  %wav,
  "ffmpeg -y -i pipe:0 -f u16le -acodec pcm_s16le -ac 2 -ar 44100 tcp://1.2.3.4:60000",
  s
)
```

2. Ideally output.external reopen_delay is fixed and LS won't retry too quickly #2756
3. Ideally I can stream the output using sockets without needing an external process / ffmpeg - see below.

# Question

@toots,

I see a new socket api was added: #2722 but it looks it was intended more for communication rather than streaming (https://github.com/savonet/liquidsoap/discussions/2703).

Could this socket api be used as an output to avoid needing an external process / ffmpeg, similar to how pipes work with output.file right now? ie

```
output.socket(
  reopen_on_error=true,
  %wav,
  "unix:/tmp/my.sock",
  s,
)
```
or
```
output.socket(
  reopen_on_error=true,
  %wav,
  "tcp://1.2.3.4:5678",
  s,
)
```

**Comment by Russsgithub** [comment]
@danbo 

Have you tried using the pipe function.

https://www.liquidsoap.info/doc-dev/reference.html#pipe

**Reply by iassenk** [reply]
After a many failed attempts, here is my working magic line. The tricky thing is that if  you specify header=true  /used by default/, then the resulting output will be a real .wav file, with all limitations - guess - 4 GB :) If, per documentation header=false,  the output will be raw audio - you should specify the input in ffmpeg:

output.external(%wav(stereo=true, channels=2, samplesize=16, header=false),
"ffmpeg -f s16le -ac 2 -i pipe:0   -acodec ..............

--------------------------------------------------------------------------------------------------------------------------------------------
Doc:
If header is false, the encoder outputs raw PCM. duration is optional and is used to set the WAV length header.

Because Liquidsoap encodes a possibly infinite stream, there is no way to know in advance the duration of encoded data. Since WAV header has to be written first, by default its length is set to the maximum possible value. If you know the expected duration of the encoded data and you actually care about the WAV length header then you should use this parameter.
---------------------------------------------------------------------------------------------------------------------------------------------

Hope this will help.
